File: src/tests/Behavioral/SliceOfArrayTypeName/SliceOfArrayTypeName.cs
namespace go;

using fmt = fmt_package;
using reflect = reflect_package;

partial class main_package {

[GoType("[3]nint")] partial struct Grid;

internal static void show(@string label, any v) {
    fmt.Printf("%-16s %%T=%-18T String()=%s\n"u8, label, v, reflect.TypeOf(v).String());
}

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly @string uint8ˢ = "[6]uint8"u8;
private static readonly @string uint8ˢ2 = "[][6]uint8"u8;
private static readonly @string intˢ = "[][3]int"u8;
private static readonly @string intˢ2 = "[2][3]int"u8;
private static readonly @string intˢ3 = "[][2][3]int"u8;
private static readonly @string map2IntIntˢ = "map[[2]int][]int"u8;
private static readonly @string gridˢ = "[]Grid"u8;
private static readonly @string byteˢ = "[]*[4]byte"u8;

internal static void Main() {
    show(uint8ˢ, new uint8[]{}.array(6));
    show(uint8ˢ2, new array<uint8>[]{new uint8[]{}.array(6)}.slice());
    show(intˢ, new array<nint>[]{new nint[]{}.array(3)}.slice());
    show(intˢ2, new array<nint>[]{}.array(2, () => new(3)));
    show(intˢ3, new array<array<nint>>[]{new array<nint>[]{new nint[]{1, 2, 3}.array(), new nint[]{4, 5, 6}.array()}.array()}.slice());
    show(map2IntIntˢ, new map<array<nint>, slice<nint>>{[new nint[]{}.array(2)] = default!});
    show(gridˢ, new Grid[]{new nint[]{}.array(3)}.slice());
    show(byteˢ, new ж<array<byte>>[]{Ꮡ(new byte[]{}.array(4))}.slice());
    fmt.Printf("slice-of-array Elem().String()=%s Len()=%d\n"u8,
        reflect.TypeOf(new array<uint8>[]{new uint8[]{}.array(6)}.slice()).Elem().String(),
        reflect.TypeOf(new array<uint8>[]{new uint8[]{}.array(6)}.slice()).Elem().Len());
}

} // end main_package

File: src/tests/Behavioral/SliceOfArrayTypeName/SliceOfArrayTypeName.go
// An ARRAY used as a container's ELEMENT must keep its length in the Go type name. A top-level
// array already did; a slice's element did not -- `[][6]uint8` rendered `[][]uint8` -- so two
// distinct Go types collapsed onto one string. That is invisible until something PRINTS the name:
// reflect's own TestDeepEqualAllocs names its subtests `ValueOf(x).Type().String()`, and the
// collapse made `t.Run` dedup the second to `#01`, so the two sides ran different-named subtests
// and the comparison reported an EMPTY verdict beside a name Go never produced.
//
// The nested rows are the point: a fix applied at one path level under-reaches, and only
// `[][2][3]int` and the map rows say whether the dims travel all the way down.
package main

import (
	"fmt"
	"reflect"
)

type Grid [3]int

func show(label string, v any) {
	fmt.Printf("%-16s %%T=%-18T String()=%s\n", label, v, reflect.TypeOf(v).String())
}

func main() {
	show("[6]uint8", [6]uint8{})
	show("[][6]uint8", [][6]uint8{{}})
	show("[][3]int", [][3]int{{}})
	show("[2][3]int", [2][3]int{})
	show("[][2][3]int", [][2][3]int{{{1, 2, 3}, {4, 5, 6}}}) // populated: the zero nested element is a routed emission gap
	show("map[[2]int][]int", map[[2]int][]int{{}: nil}) // present entry: an EMPTY map is increment B's stated boundary
	show("[]Grid", []Grid{{}})
	show("[]*[4]byte", []*[4]byte{&[4]byte{}}) // explicit &: the elided form trips a converter gap, routed

	// Elem() is the type-side question the value-side fix cannot answer on its own.
	fmt.Printf("slice-of-array Elem().String()=%s Len()=%d\n",
		reflect.TypeOf([][6]uint8{{}}).Elem().String(), // present element: the empty literal is the stated boundary
		reflect.TypeOf([][6]uint8{{}}).Elem().Len())
}
